source/common/protobuf/deterministic_hash.cc

```diff
@@ -0,0 +1,239 @@
+#if defined(ENVOY_ENABLE_FULL_PROTOS)
+#include "source/common/protobuf/deterministic_hash.h"
+
+#include "source/common/common/assert.h"
+#include "source/common/common/hash.h"
+
+namespace Envoy {
+namespace DeterministicProtoHash {
+namespace {
+
+#define REFLECTION_FOR_EACH(get_type, value_type)                                                  \
+  if (field->is_repeated()) {                                                                      \
+    for (const auto q : reflection->GetRepeatedFieldRef<value_type>(message, field)) {             \
+      seed = HashUtil::xxHash64(absl::string_view{reinterpret_cast<const char*>(&q), sizeof(q)},   \
+                                seed);                                                             \
+    }                                                                                              \
+  } else {                                                                                         \
+    const auto q = reflection->Get##get_type(message, field);                                      \
+    seed =                                                                                         \
+        HashUtil::xxHash64(absl::string_view{reinterpret_cast<const char*>(&q), sizeof(q)}, seed); \
+  }
+
+uint64_t reflectionHashMessage(const Protobuf::Message& message, uint64_t seed = 0);
+uint64_t reflectionHashField(const Protobuf::Message& message,
+                             const Protobuf::FieldDescriptor* field, uint64_t seed);
+
+struct MapFieldComparer {
+  MapFieldComparer(const Protobuf::Message& first_msg)
+      : reflection_(*first_msg.GetReflection()), descriptor_(*first_msg.GetDescriptor()),
+        key_field_(*descriptor_.map_key()), compare_fn_(selectCompareFn()) {}
+  bool operator()(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return (this->*compare_fn_)(a, b);
+  }
+
+private:
+  bool compareByInt32(const Protobuf::Message& a, const Protobuf::Message& b) {
+    return reflection_.GetInt32(a, &key_field_) < reflection_.GetInt32(b, &key_field_);
```

Comment:
Yes, I was a bit tempted to make a more generic function for this called something like `protoReflectionGetThatShouldBeInTheDamnLibrary<T>` so that here it could be templated (and in the macro too rather than tokenized).